### PR TITLE
[TEAM1] Data logging

### PR DESCRIPTION
**Pull request by JaafarRammal**

### CODE REVIEW

File: 003_data_logging/003_data_logging.tex [hunk 1]
```diff
@@ -0,0 +1,86 @@
+\chapter{Data Logging and Dashboard}\label{data_logging}
+
+\section{Why do we need data?}
+
+Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc... Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available).
```
Comment (daryllimyt) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
```

File: 003_data_logging/003_data_logging.tex [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+"level":"info","msg":"Reporting agent state of team 5 agent","reporter":"agent"}
+
+{"Floor:":7,"HP:":100,"Social motive:":"Altruist","agent_id":"2bfd3cfb",
+"agent_type":"Team6","level":"info","msg":"Reporting agent state:","reporter":"agent"}
+
+{"agent_id":"7873bb50","agent_type":"Team1","floor":5,"health":100,
+"level":"info","msg":"Reporting agent state","reporter":"agent"}
+\end{verbatim}
+
+The general log dumps are not structured yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.
```
Comment (daryllimyt) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The general log dumps are unstructured - yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.
```

File: 003_data_logging/003_data_logging.tex [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+\section{Structured logs}
+
+The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. Following consultations with the agent teams and relating as fit to the problem, the following structured logs where included:
```
Comment (daryllimyt) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. After discussions with the agent teams and relating as fit to the problem, the following structured logs where included:
```

File: 003_data_logging/003_data_logging.tex [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    \item \textbf{Food}: food available on the platform and food taken by agents.
+    \item \textbf{Death}: agent death (including agent state at death time)
+    \item \textbf{Messages}: receiver, sender, message type and content
+    \item \textbf{Utility}: agent utility (including agent state)
+    \item \textbf{Story}: a combination of the above logs in a sequential manner
+\end{enumerate}
+
+The agent state encapsulates by default the agent age, floor, HP, utility, and custom state (e.g. emotions or memory).
+
+In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which would be required for a visualisation dashboard). Structured logs have a well defined schema which does not depend on custom agents. For example, a food log will always have this form:
```
Comment (daryllimyt) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which would be required for a visualisation dashboard). Structured logs have a well defined schema which does not depend on custom agents. For example, a food log will always take the following form:
```

File: 003_data_logging/003_data_logging.tex [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+\begin{verbatim}
+{"day":2,"food":19,"level":"info","tick":222,"time":"2022-01-17T02:13:37Z"}
+{"day":2,"food":0,"level":"info","tick":262,"time":"2022-01-17T02:13:37Z"}
+{"day":3,"food":100,"level":"info","tick":321,"time":"2022-01-17T02:13:37Z"}
+{"day":3,"food":90,"level":"info","tick":322,"time":"2022-01-17T02:13:37Z"}
+\end{verbatim}
+
+With structured logs setup, it is now possible to bundle logging into an interface and connect it to a visualisation tool.
```
Comment (daryllimyt) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
With structured logs set up, it is now possible to bundle logging into an interface and connect it to a visualisation tool.
```

File: 003_data_logging/003_data_logging.tex [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+\section{Logging interface}
+
+It was important to setup logging such as structured loggers could be easily extended with new loggers, while preserving the possibility to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.
```
Comment (daryllimyt) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
It was important to set up logging such as structured loggers could be easily extended with new loggers, while preserving the ability to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.
```

File: 003_data_logging/003_data_logging.tex [hunk 7]
```diff
@@ -0,0 +1,86 @@
+\chapter{Data Logging and Dashboard}\label{data_logging}
+
+\section{Why do we need data?}
```
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
\section{Need for Data}
```

Probably best to keep it formal

**Review by daryllimyt — CHANGES_REQUESTED** [review]

File: 003_data_logging/003_data_logging.tex (hunk 1 above)
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
Data logging allows for the analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question, such as messaging frequency, food variation over time or the agents' death rate, with analysis also requiring the linking of different simulation parameters with their logged data (e.g., the number of agents or the amount of food available). 
```

I think this flows better

File: 003_data_logging/003_data_logging.tex [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc... Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
+
+It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parametrisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.
+
+\section{Parametrisation}
+
+First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment. A configuration object would therefore be passed to the simulation environment instance, and it would define the following simulation parameters:
```
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment and a configuration object would subsequently be passed to the simulation environment instance to define the following simulation parameters:
```

Again, let me know if this is better. Prevents the staggered sentences and repetition of 'therefore'

File: 003_data_logging/003_data_logging.tex [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        \item Total days
+        \item Reshuffle period
+    \end{enumerate}
+\end{multicols}
+
+The configuration parameters could then be written in a simple \texttt{json} file which would be parsed into a configuration object.
+
+\section{General log dumps}
+
+Following the configuration work, a general logger was setup. The general logger is basically a single shared logger between the different instances in the system (simulation, tower, agents, etc...). Each instance appends by default an identifier to a log line in the general logger log file (named \texttt{main.json}). Here is an example below:
```
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
Following the configuration work, a general logger was set up. The general logger is effectively a single shared logger between the different instances in the system (simulation, tower, agents, etc...). Each instance appends by default an identifier to a log line in the general logger log file (named \texttt{main.json}). Here is an example below:
```

File: 003_data_logging/003_data_logging.tex (hunk 2 above)
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The general log dumps are not structured, yet they enable agent teams to directly log custom states from their agents and use them, for example, in evolutionary algorithms or machine learning training.
```

File: 003_data_logging/003_data_logging.tex (hunk 3 above)
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. Following consultations with the agent teams and relating as fit to the problem, the following structured logs were included:
```

File: 003_data_logging/003_data_logging.tex [hunk 10]
```diff
@@ -0,0 +1,86 @@
+\chapter{Data Logging and Dashboard}\label{data_logging}
+
+\section{Why do we need data?}
+
+Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
+
+It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parametrisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.
+
+\section{Parametrisation}
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
\section{Parameterisation}
```

File: 003_data_logging/003_data_logging.tex (hunk 7 above)
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
\section{Why Do We Need Data?}
```

File: 003_data_logging/003_data_logging.tex [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+\begin{enumerate}
+    \item \textbf{Food}: food available on the platform and food taken by agents.
+    \item \textbf{Death}: agent death (including agent state at death time)
+    \item \textbf{Messages}: receiver, sender, message type and content
+    \item \textbf{Utility}: agent utility (including agent state)
+    \item \textbf{Story}: a combination of the above logs in a sequential manner
+\end{enumerate}
+
+The agent state encapsulates by default the agent age, floor, HP, utility, and custom state (e.g. emotions or memory).
```
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The agent state encapsulates by default the agent's age, floor, HP, utility, and custom state (e.g., emotions or memory).
```

File: 003_data_logging/003_data_logging.tex (hunk 6 above)
Comment (MattSScott) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
It was important to set up logging such that structured loggers could be easily extended with new loggers, while preserving the possibility to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.
```

**Review by daryllimyt — DISMISSED** [review]

**Review by MattSScott — DISMISSED** [review]
Some grammar

File: 003_data_logging/003_data_logging.tex [hunk 12]
```diff
@@ -0,0 +1,86 @@
+\chapter{Data Logging and Dashboard}\label{data_logging}
+
+\section{Why do we need data?}
+
+Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available).
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messaging frequency, food variation over time, agent death rates, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
```

File: 003_data_logging/003_data_logging.tex [hunk 13]
```diff
@@ -0,0 +1,86 @@
+\chapter{Data Logging and Dashboard}\label{data_logging}
+
+\section{Why do we need data?}
+
+Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
+
+It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parametrisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parameterisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.
```

File: 003_data_logging/003_data_logging.tex (hunk 11 above)
Comment (JaafarRammal) [reply 003_data_logging/003_data_logging.tex]: Do we need the comma in `(e.g., emotions or memory).` or typo?

File: 003_data_logging/003_data_logging.tex [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+\section{Parametrisation}
+
+First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment. A configuration object would therefore be passed to the simulation environment instance, and it would define the following simulation parameters:
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
First, it was important to enable an effective parameterisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment. A configuration object is passed to the simulation environment instance, and defines the following simulation parameters:
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: 003_data_logging/003_data_logging.tex [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+\section{Need for Data}
+
+Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
+
+It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parametrisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.
+
+\section{Parametrisation}
+
+First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment and a configuration object would subsequently be passed to the simulation environment instance to define the following simulation parameters:
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
First, it was important to enable an effective parameterisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment and a configuration object is passed to the simulation environment instance to define the following simulation parameters:
```

File: 003_data_logging/003_data_logging.tex [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        \item Agents per floor
+        \item Amount of food on the platform
+        \item Food per agent ratio (optional)
+        \item Ticks per floor
+        \item Total days
+        \item Reshuffle period
+    \end{enumerate}
+\end{multicols}
+
+The configuration parameters could then be written in a simple \texttt{json} file which would be parsed into a configuration object.
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The configuration parameters are then written in a JSON file which can be parsed into a configuration object.
```

File: 003_data_logging/003_data_logging.tex [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+\section{General log dumps}
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
\section{General Log Dumps}
```

File: 003_data_logging/003_data_logging.tex (hunk 11 above)
Comment (MattSScott) [reply 003_data_logging/003_data_logging.tex]: we do - agreed on slack last night

File: 003_data_logging/003_data_logging.tex [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+"level":"info","msg":"Reporting agent state of team 5 agent","reporter":"agent"}
+
+{"Floor:":7,"HP:":100,"Social motive:":"Altruist","agent_id":"2bfd3cfb",
+"agent_type":"Team6","level":"info","msg":"Reporting agent state:","reporter":"agent"}
+
+{"agent_id":"7873bb50","agent_type":"Team1","floor":5,"health":100,
+"level":"info","msg":"Reporting agent state","reporter":"agent"}
+\end{verbatim}
+
+The general log dumps are unstructured - yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The general log dumps are unstructured, but enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.
```

File: 003_data_logging/003_data_logging.tex [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+\section{Structured logs}
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
\section{Structured Logs}
```

File: 003_data_logging/003_data_logging.tex [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+
+The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. After discussions with the agent teams and relating as fit to the problem, the following structured logs where included:
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data are captured optimally to generate lightweight, understandable logs. After discussions with agent teams, the following structured logs where included:
```

File: 003_data_logging/003_data_logging.tex [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    \item \textbf{Food}: food available on the platform and food taken by agents.
+    \item \textbf{Death}: agent death (including agent state at death time)
+    \item \textbf{Messages}: receiver, sender, message type and content
+    \item \textbf{Utility}: agent utility (including agent state)
+    \item \textbf{Story}: a combination of the above logs in a sequential manner
+\end{enumerate}
+
+The agent state encapsulates by default the agent age, floor, HP, utility, and custom state (e.g. emotions or memory).
+
+In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which would be required for a visualisation dashboard). Structured logs have a well defined schema which does not depend on custom agents. For example, a food log will always take the following form:
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which is required for a visualisation dashboard). Structured logs have a well-defined schema which does not depend on custom agents. For example, a food log will always take the following form:
```

File: 003_data_logging/003_data_logging.tex [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+\begin{verbatim}
+{"day":2,"food":19,"level":"info","tick":222,"time":"2022-01-17T02:13:37Z"}
+{"day":2,"food":0,"level":"info","tick":262,"time":"2022-01-17T02:13:37Z"}
+{"day":3,"food":100,"level":"info","tick":321,"time":"2022-01-17T02:13:37Z"}
+{"day":3,"food":90,"level":"info","tick":322,"time":"2022-01-17T02:13:37Z"}
+\end{verbatim}
+
+With structured logs set up, it is now possible to bundle logging into an interface and connect it to a visualisation tool.
+
+\section{Logging interface}
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
\section{Logging Interface}
```

File: 003_data_logging/003_data_logging.tex [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+It was important to set up logging such as structured loggers could be easily extended with new loggers, while preserving the ability to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
It was important to set up logging such that structured loggers could be easily extended with new loggers, while preserving the ability to generate general logs. The following design was adopted to encapsulate the simulation configuration and the loggers.
```

File: 003_data_logging/003_data_logging.tex [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+\begin{figure}[htb]
+    \centering
+    \includegraphics[width=0.8\linewidth]{003_data_logging/images/struct.png}
+    \caption{Diagram showing how logging fits in the system}
+    \label{fig:design_logging}
+\end{figure}
+
+The entry point to the simulation dispatcher was then linked to both the CLI as well as a server (with HTTP listeners) to support a browser-based dashboard.
```
Comment (jzzheng22) [inline 003_data_logging/003_data_logging.tex]:
```suggestion
The entry point to the simulation dispatcher is linked to both the command line interface as well as a server (with HTTP listeners) to support a browser-based dashboard.
```

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

**Review by daryllimyt — DISMISSED** [review]
LGTS (Lets go to sleep)

**Review by jzzheng22 — DISMISSED** [review]
why are we still doing this

**Review by jzzheng22 — APPROVED** [review]
help

**Review by daryllimyt — APPROVED** [review]

**Pull request merged**